src/Messages/OWSMessageSender.m

```diff
@@ -281,9 +296,14 @@ - (void)tryWithRemainingRetries:(NSUInteger)remainingRetries
 {
     DDLogDebug(@"%@ remainingRetries: %lu", self.tag, (unsigned long)remainingRetries);
 
+    // Use this flag to ensure a given operation only succeeds or fails once.
+    __block BOOL onceFlag = NO;
     RetryableFailureHandler retryableFailureHandler = ^(NSError *_Nonnull error) {
         DDLogInfo(@"%@ Sending failed.", self.tag);
 
+        OWSAssert(!onceFlag);
```

Comment:
Not quite, and:

* There's a lot of complexity in the message sending process. 
* This is core functionality that we want to have high confidence in.
* This flag is cheap and local, so why not.